File: sprint_3/misc/merge_sort.py
"""
Merge sort.
Time: O(n log n). Space: O(n)
"""

from dataclasses import dataclass


def merge_sort(arr: list[int]) -> list[int]:
    if len(arr) <= 1:
        return arr
    left = merge_sort(arr[0 : len(arr) // 2])
    right = merge_sort(arr[len(arr) // 2 : len(arr)])
    result = [0] * len(arr)
    l, r, k = 0, 0, 0
    while l < len(left) and r < len(right):
        if left[l] <= right[r]:
            result[k] = left[l]
            l += 1
        else:
            result[k] = right[r]
            r += 1
        k += 1

    while l < len(left):
        result[k] = left[l]
        l += 1
        k += 1
    while r < len(right):
        result[k] = right[r]
        r += 1
        k += 1

    return result


@dataclass
class TestData:
    arr: list[int]
    expected: list[int]


def test():
    tds = [
        TestData([], []),
        TestData([-1, -4, 0, -2], [-4, -2, -1, 0]),
        TestData([89, 4, 11, -8, 14], [-8, 4, 11, 14, 89]),
        TestData([4, 89, 11, -8, 14, -22, -8, 0], [-22, -8, -8, 0, 4, 11, 14, 89]),
    ]
    for td in tds:
        got = merge_sort(td.arr)
        assert got == td.expected, f"expected {td.expected}, got {got}, input {td.arr}"


if __name__ == "__main__":
    test()

File: sprint_3/misc/merge_sort.c
/*
Merge sort.
Time: O(n log n). Space: O(n)
*/

#include <stdio.h>
#include <stdlib.h>

void print_array(int *array, int len, const char *msg_prefix);
void merge(int *array, int *workspace, size_t start, size_t end, size_t mid);
void merge_sort(int *array, int *workspace, size_t start, size_t end);

int main(void) {
    int arr[] = {-1, -8, 0, 44, 12, -41, 21, 14, 99, -909};
    size_t arr_len = sizeof(arr) / sizeof(int);
    print_array(arr, arr_len, "original array: ");
    int workspace[arr_len];
    merge_sort(arr, workspace, 0, arr_len);
    print_array(arr, arr_len, "sorted array: ");
    return 0;
}

void print_array(int *array, int len, const char *msg_prefix) {
    printf("%s", msg_prefix);
    for (int i = 0; i < len; i++) {
        printf("%d ", array[i]);
    }
    printf(" \n");
}

void merge_sort(int *array, int *workspace, size_t start, size_t end) {
    if (end - start <= 1) {
        return;
    }

    size_t mid = start + (end - start) / 2;
    merge_sort(array, workspace, start, mid);
    merge_sort(array, workspace, mid, end);

    merge(array, workspace, start, end, mid);
}

void merge(int *array, int *workspace, size_t start, size_t end, size_t mid) {
    size_t i = start, j = mid, k = 0;

    while (i < mid && j < end) {
        if (array[i] <= array[j]) {
            workspace[k++] = array[i++];
        } else {
            workspace[k++] = array[j++];
        }
    }
    while (i < mid) {
        workspace[k++] = array[i++];
    }
    while (j < end) {
        workspace[k++] = array[j++];
    }
    for (i = start, k = 0; i < end; i++, k++) {
        array[i] = workspace[k];
    }
}
